### PR TITLE
Add documentation note on All and superclasses - closes #66

### LINKED ISSUE #66: Super classes and All - improve documentation
I've documented an issue I've encountered using the `All` constraint on [stack overflow](https://stackoverflow.com/questions/50777865/super-classes-with-all-from-generics-sop) but I'm wondering if this is a better place, so I figured I'd drop an issue here as well - if the SO gets answered then I'll close this issue with a reference to it.

### PR DESCRIPTION
**Pull request by matchwood**
Closes #66

### CODE REVIEW

File: src/Generics/SOP/Constraint.hs [hunk 1]
```diff
@@ -42,6 +42,24 @@ import Generics.SOP.Sing
 --
 -- means that 'f' can assume that all elements of the n-ary
 -- product satisfy 'Eq'.
+
```
Comment (phadej) [inline src/Generics/SOP/Constraint.hs]: The empty line breaks haddock

File: src/Generics/SOP/Constraint.hs [hunk 2]
```diff
@@ -42,6 +42,24 @@ import Generics.SOP.Sing
...
+-- Note on superclasses: ghc cannot deduce superclasses from 'All'
+-- constraints.
+-- You might expect the following to compile
+-- > class (Eq a) => MyClass a
```
Comment (phadej) [inline src/Generics/SOP/Constraint.hs]: there must be an empty line (with `--` though) to split the paragraph from code block

File: src/Generics/SOP/Constraint.hs [hunk 3]
```diff
@@ -42,6 +42,24 @@ import Generics.SOP.Sing
...
+-- >
+-- > foo :: (All Eq xs) => NP f xs -> z
+-- > foo = [..]
+-- >
+-- > bar :: (All MyClass xs) => NP f xs -> x
+-- > bar = foo
+-- but it will fail with an error saying that it was unable to
+-- deduce the class constraint 'AllF Eq xs' (or similar) in the
+-- definition of 'bar'.
+-- In cases like this you can use 'Dict' from 'Generics.SOP.Dict'
```
Comment (phadej) [inline src/Generics/SOP/Constraint.hs]: modules are references with double quotes: `"Generics.Sop.Dict"`

File: src/Generics/SOP/Constraint.hs [hunk 4]
```diff
@@ -42,6 +42,24 @@ import Generics.SOP.Sing
...
+-- You might expect the following to compile
+-- > class (Eq a) => MyClass a
+-- >
+-- > foo :: (All Eq xs) => NP f xs -> z
+-- > foo = [..]
+-- >
+-- > bar :: (All MyClass xs) => NP f xs -> x
+-- > bar = foo
+-- but it will fail with an error saying that it was unable to
+-- deduce the class constraint 'AllF Eq xs' (or similar) in the
```
Comment (phadej) [inline src/Generics/SOP/Constraint.hs]: Renders as 

screenshot from 2018-08-08 13-37-01
---

Use 

```
@'AllF' 'Eq' xs@
```

which renders as

screenshot from 2018-08-08 13-38-39

File: src/Generics/SOP/Constraint.hs (hunk 3 above)
Comment (phadej) [reply src/Generics/SOP/Constraint.hs]: Writing full paths for `Dict` makes it a link too

```
-- In cases like this you can use 'Generics.SOP.Dict.Dict' from "Generics.SOP.Dict"
```

screenshot from 2018-08-08 13-41-41

File: src/Generics/SOP/Constraint.hs (hunk 1 above)
Comment (matchwood) [reply src/Generics/SOP/Constraint.hs]: Many apologies, I don't have haddocks generation as part of my workflow. I'll set it up now and update the PR when I've checked my new changes.

**Comment by kosmikus** [comment]
Thanks for the PR. I have restarted CI and will try to look at this once I'm back from vacation.

**Comment by kosmikus** [comment]
I'm going to merge this, as it's certainly an improvement over the status quo. Thanks a lot for the contribution! This being said, I don't find linking to SO questions ideal. I'd prefer providing more detailed info directly within the library. I think the `Dict` module probably deserves some introduction with example use cases and explaining what it does. We could then link to that from `All`.

**Pull request merged**